portal/portal-service-impl/impl/src/java/org/sakaiproject/portal/service/BullhornServiceImpl.java

```diff
@@ -0,0 +1,388 @@
+                                            String url = assignmentService.getDeepLink(siteId, assignmentId);
+                                            doAcademicInsert(from, to, event, ref, title, siteId, e.getEventTime(), url);
+                                            countCache.remove(to);
+                                        }
+                                    }
+                                }
+                            } catch (IdUnusedException idue) {
+                                log.error("Failed to find either the assignment or the site", idue);
+                            }
+                        } else if (AssignmentConstants.EVENT_GRADE_ASSIGNMENT_SUBMISSION.equals(event)) {
+                            String siteId = pathParts[3];
+                            String submissionId = pathParts[pathParts.length - 1];
+                            try {
+                                SecurityAdvisor sa = unlock();
+                                AssignmentSubmission submission = assignmentService.getSubmission(submissionId);
+                                lock(sa);
+                                if (submission.getGradeReleased()) {
+                                    Site site = siteService.getSite(siteId);
+                                    Assignment assignment = submission.getAssignment();
+                                    String title = assignment.getTitle();
+                                    for (String to : ((List<String>) submission.getSubmitterIds())) {
+                                        switchUser(to);
+                                        String url = assignmentService.getDeepLink(siteId, assignment.getId());
+                                        doAcademicInsert(from, to, event, ref, title, siteId, e.getEventTime(), url);
+                                        countCache.remove(to);
+                                    }
+                                }
+                            } catch (IdUnusedException idue) {
+                                log.error("Failed to find either the submission or the site", idue);
+                            }
+                        }
+                    }
+                } catch (Exception ex) {
+                    log.error("Caught exception whilst handling events", ex);
+                }
+            }).start();
+        }
+    }
+
+    private SecurityAdvisor unlock() {
+
+        SecurityAdvisor securityAdvisor = new SecurityAdvisor() {
+                public SecurityAdvice isAllowed(String userId, String function, String reference) {
+                    return SecurityAdvice.ALLOWED;
+                }
+            };
+        securityService.pushAdvisor(securityAdvisor);
+        return securityAdvisor;
+    }
+
+    private void lock(SecurityAdvisor securityAdvisor) {
+        securityService.popAdvisor(securityAdvisor);
+    }
+
+    private String switchUser(String to) {
+
```

Comment:
Yeah, I'd agree with this, having exception unprotected user switching in a potential multithreaded area of code seems like it could result in some hard to track issues. I don't feel like there's any other code doing this. If it's because of security probably better to use the securityAdvisors or internal methods that don't do security checks rather than changing the session.